### PR TITLE
Implement Azure IaC generator with Bicep and diagram support

### PR DESCRIPTION
**Pull request by Ba4bes**
## Description

Created to replace #1892 was messed up with some creative use of git. 

 Adds azure-iac-generator, to create bicep templates based on either draw.io diagrams or based on the live Azure environment

## Checklist

- [ ] Tests pass locally (`cd tests && npm test`)
- [ ] **If modifying skill descriptions:** verified routing correctness with integration tests (`npm run test:skills:integration -- <skill>`)
- [ ] **If modifying skill `USE FOR` / `DO NOT USE FOR` / `PREFER OVER` clauses:** confirmed no routing regressions for competing skills
- [ ] Version bumped in skill frontmatter (if skill files changed)

### CODE REVIEW

**Comment by Ba4bes** [comment]
@jongio @tmeschter  And... I did it again. Created this to replace #1892 

Previous comments as far as I can see were resolved, found a few bugs myself and will wait for copilot for extra feedback :)

File: plugin/skills/azure-iac-generator/references/auto-detection-rules.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+| Condition | Auto-Setting |
+|-----------|-------------|
+| Resource has a Private Endpoint connection | Set `publicNetworkAccess: "Disabled"` on target resource |
+| App Service connected to a Subnet | Set `vnetIntegrationSubnet` to the subnet reference |
+| Private Endpoint connected to SQL Server | Set `groupIds: ["sqlServer"]` |
+| Private Endpoint connected to Storage Account | Set `groupIds: ["blob"]` |
+| Private Endpoint connected to App Service | Set `groupIds: ["sites"]` |
+| Private Endpoint connected to Key Vault | Set `groupIds: ["vault"]` |
+| Private Endpoint connected to Cosmos DB | Set `groupIds: ["Sql"]` |
+| Private Endpoint exists in a subnet | Set `privateEndpointNetworkPolicies: "Disabled"` on that subnet |
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/references/auto-detection-rules.md]: fixed as suggested

File: plugin/skills/azure-iac-generator/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
+---
+name: azure-iac-generator
+description: "Generate deployment-ready Bicep templates from existing Azure environments or Draw.io architecture diagrams. Reverse-engineer live infrastructure into Infrastructure as Code. WHEN: generate bicep, azure to bicep, generate bicep from azure, bicep from diagram, diagram to bicep, create bicep templates from resources, export infrastructure as code, generate infrastructure code, reverse engineer azure, generate iac from azure. DO NOT USE FOR: creating new applications for Azure (use azure-prepare), deploying existing Bicep (use azure-validate then azure-deploy), comparing Bicep against live Azure (use azure-infrastructure-sync)."
+license: MIT
+metadata:
+  author: Microsoft
+  version: "0.0.0-placeholder"
+---
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/SKILL.md:8]: added version.json file

File: plugin/skills/azure-iac-generator/references/version-currency.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+Always note the version choice in `.bicepparam` comments. Treat the example below as a template only: look up the current supported runtime and support window at generation time rather than copying literal versions or dates from this document.  
+***bicep
+// Runtime stack for the App Service.
+// Replace the placeholders below with the current supported values at generation time.
+// → (, supported until )
+// → (, supported until )
+// → (, supported until )
+// → (, supported until )
+param appServiceRuntimeStack = ''
+```
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/references/version-currency.md:31]: fixed to backtics

File: tests/azure-iac-generator/integration.test.ts [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            followUp: ["Continue until all files are written."],
+          });
+
+          expect(workspacePath).toBeDefined();
+          const hasMainBicep = doesWorkspaceFileIncludePattern(
+            workspacePath!,
+            /targetScope|module\s+\w+|param\s+\w+/,
+            /main\.bicep$/
+          );
+          expect(hasMainBicep).toBe(true);
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply tests/azure-iac-generator/integration.test.ts:184]: updated as suggested

**Comment by Ba4bes** [comment]
Fixed some incosistencies in the skill-files as well

**Review by jongio — COMMENTED** [review]
1. Frontmatter version is `0.0.0-placeholder` - should be `1.0.0` per repo conventions for new skills.
2. Stencil mapping JSON is missing 7 resource types that the resource model lists. Diagrams with those icons won't parse correctly.
3. `version.json` isn't referenced by any build tooling in this repo and can be removed.
4. Trigger tests need boundary prompts that route to other skills to catch routing conflicts.

`azure-authentication.md` and `bicep-parsing.md` are missing trailing newlines.

File: plugin/skills/azure-iac-generator/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+name: azure-iac-generator
+description: "Generate deployment-ready Bicep templates from existing Azure environments or Draw.io architecture diagrams. Reverse-engineer live infrastructure into Infrastructure as Code. WHEN: generate bicep, azure to bicep, generate bicep from azure, bicep from diagram, diagram to bicep, create bicep templates from resources, export infrastructure as code, generate infrastructure code, reverse engineer azure, generate iac from azure. DO NOT USE FOR: creating new applications, deploying existing resources, comparing environments."
+license: MIT
+metadata:
+  author: Microsoft
+  version: "0.0.0-placeholder"
```
Comment (jongio) [inline plugin/skills/azure-iac-generator/SKILL.md:7]: `0.0.0-placeholder` doesn't match repo conventions. Every other skill uses an actual semver version, and the [instructions](.github/instructions/skill-files.instructions.md) say "Set to `1.0.0` for new skills."

```suggestion
  version: "1.0.0"
```

File: plugin/skills/azure-iac-generator/references/azure-stencil-mapping.json [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    },
+    {
+      "type": "Microsoft.OperationalInsights/workspaces",
+      "imagePath": "img/lib/azure2/monitor/Log_Analytics_Workspaces.svg"
+    },
+    {
+      "type": "Microsoft.ManagedIdentity/userAssignedIdentities",
+      "imagePath": "img/lib/azure2/identity/Managed_Identities.svg"
+    }
+  ]
```
Comment (jongio) [inline plugin/skills/azure-iac-generator/references/azure-stencil-mapping.json:179]: The mapping covers 25 resource types but the resource model doc lists 32. Missing from the mapping: `Microsoft.Web/sites/functions` (Azure Functions), `Microsoft.Network/azureFirewalls`, `Microsoft.Network/bastionHosts`, `Microsoft.Network/virtualNetworkGateways`, `Microsoft.ApiManagement/service`, `Microsoft.SignalRService/signalR`, `Microsoft.CognitiveServices/accounts`. Diagrams using these icons won't be recognized by the parser and the resources will silently drop from generated Bicep.

File: tests/azure-iac-generator/triggers.test.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    test.each(shouldTriggerPrompts)(
+      'triggers on: "%s"',
+      (prompt) => {
+        const result = triggerMatcher.shouldTrigger(prompt);
+        expect(result.triggered).toBe(true);
+      }
+    );
+  });
+
+  describe("Should NOT Trigger", () => {
```
Comment (jongio) [inline tests/azure-iac-generator/triggers.test.ts:56]: The negative cases here only test clearly unrelated prompts (weather, poems). Consider adding boundary cases that should route to *other* Azure skills instead:
- `Deploy my Bicep templates to Azure` (azure-deploy territory)
- `Create a new web app with Bicep` (azure-prepare territory)
- `Compare my dev and prod Azure environments` (listed in DO NOT USE FOR but not tested as a negative trigger)

These catch routing conflicts if description keywords overlap with other skills.

File: plugin/skills/azure-iac-generator/version.json [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{
+  "version": "1.1",
```
Comment (jongio) [inline plugin/skills/azure-iac-generator/version.json]: This file isn't referenced by any build tooling in the repo, and no other skill has a `version.json`. The copilot reviewer suggested adding it, but the assumption about gulpfile/NBGV version stamping doesn't apply here. Consider removing to avoid dead config.

File: plugin/skills/azure-iac-generator/SKILL.md (hunk 5 above)
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/SKILL.md:7]: I find the information a bit confusing in the instructions, as these two lines contradict each other. I think it was actually copilot who requested the current format:
```Semver format (`X.Y.Z`). Set to `"1.0.0"` for new skills. For skills under `plugin/`, versions are stamped automatically at build time by NBGV — use `"0.0.0-placeholder"` in source.```

I changed it back to 1.0.0 now

File: plugin/skills/azure-iac-generator/references/azure-stencil-mapping.json (hunk 6 above)
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/references/azure-stencil-mapping.json:179]: Added the extra resource types and added extra instructions to diagram-parsing.md to make it clear what should be done when the resource is not in this mapping.

File: plugin/skills/azure-iac-generator/version.json (hunk 8 above)
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/version.json]: I applied the Copilot review because I do see version.json under a lot of the other skills, ``plugin\skills\azure-cost\version.json` for example. Removed the file now

File: tests/azure-iac-generator/triggers.test.ts (hunk 7 above)
Comment (Ba4bes) [reply tests/azure-iac-generator/triggers.test.ts:56]: I added extra cases, although I needed to keep them a bit more generic because the tests will fail if terms like "bicep"  or "azure" are used.

File: plugin/skills/azure-iac-generator/references/auto-detection-rules.md [hunk 9]
```diff
@@ -0,0 +1,18 @@
+# Auto-Detection Rules
+
+Settings automatically applied based on diagram topology during Bicep generation.
+
+| Condition | Auto-Setting |
+|-----------|-------------|
+| Resource has a Private Endpoint connection | Set `publicNetworkAccess: 'Disabled'` on target resource |  
+| App Service connected to a Subnet | Set `vnetIntegrationSubnet` to the subnet reference |  
+| Private Endpoint connected to SQL Server | Set `groupIds: ['sqlServer']` |  
+| Private Endpoint connected to Storage Account | Set `groupIds: ['blob']` |
```
Comment (jongio) [inline plugin/skills/azure-iac-generator/references/auto-detection-rules.md]: Storage accounts support multiple private endpoint group IDs: `blob`, `file`, `queue`, `table`, `web`, `dfs`. Hardcoding `['blob']` here means any PE targeting file shares, queues, or tables will get incorrect Bicep. Consider making this conditional on the connection target, or at minimum document this as a blob-only default.

File: plugin/skills/azure-iac-generator/references/procedures/diagram-parsing.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+5. **Extract connections**: For each `mxCell` with `edge="1"`:
+   - Extract `source` and `target` cell IDs
+   - Determine relationship type from edge style:
+
+   | Edge Style | Relationship |
+   |---|---|
+   | `strokeColor=#0078D4` | `connects` (data flow) |
+   | `strokeColor=#E81123` | `secures` (security / private link) |
+   | `strokeColor=#999999` + `dashed=1` | `depends` (dependency) |
+   | `strokeColor=#00A4EF` + `dashed=1` | `peers` (network link) |
```
Comment (jongio) [inline plugin/skills/azure-iac-generator/references/procedures/diagram-parsing.md:46]: This color-based mapping works when users follow the convention, but most Draw.io edges use the default style with no explicit `strokeColor`. What relationship type gets assigned when an edge has no matching color? A default (e.g., fall back to `connects` for unmatched edges) would prevent silent drops or errors on real-world diagrams.

File: plugin/skills/azure-iac-generator/references/resource-configs-compute.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+| runtimeStack | `properties.siteConfig.linuxFxVersion` or `windowsFxVersion` | Format: `RUNTIME\|VERSION` |
+| httpsOnly | `properties.httpsOnly` | |
+| minTlsVersion | `properties.siteConfig.minTlsVersion` | config sub-call |
+| alwaysOn | `properties.siteConfig.alwaysOn` | config sub-call |
+| http20Enabled | `properties.siteConfig.http20Enabled` | config sub-call |
+| ftpsState | `properties.siteConfig.ftpsState` | config sub-call |
+| publicNetworkAccess | `properties.publicNetworkAccess` | |
+| vnetIntegrationSubnet | `properties.virtualNetworkSubnetId` | |
+
+### Microsoft.Web/sites/functions
```
Comment (jongio) [inline plugin/skills/azure-iac-generator/references/resource-configs-compute.md]: Type mismatch: `Microsoft.Web/sites/functions` in ARM refers to an individual function within a function app (a child resource). The function app itself is `Microsoft.Web/sites` with `kind` containing `functionapp` - which the stencil mapping represents as `Microsoft.Web/sites[functionapp]`. The `az functionapp show` fallback is correct for the app-level resource, but this heading should align with the stencil mapping's discriminator pattern to avoid confusion during diagram-to-model resolution.

**Review by jongio — COMMENTED** [review]
Three items the existing reviews didn't cover - all related to correctness of generated Bicep when the skill runs.

1. Storage private endpoint group IDs default to blob only - won't produce correct Bicep for file/queue/table endpoints
2. Diagram edge parsing has no fallback when edges don't carry an explicit stroke color
3. Function App type key mismatch between resource-configs-compute.md and the stencil mapping

File: plugin/skills/azure-iac-generator/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
+---
+name: azure-iac-generator
+description: "Generate deployment-ready Bicep templates from existing Azure environments or Draw.io architecture diagrams. Reverse-engineer live infrastructure into Infrastructure as Code. WHEN: generate bicep, azure to bicep, generate bicep from azure, bicep from diagram, diagram to bicep, create bicep templates from resources, export infrastructure as code, generate infrastructure code, reverse engineer azure, generate iac from azure. DO NOT USE FOR: creating new applications, deploying existing resources, comparing environments."
+license: MIT
+metadata:
+  author: Microsoft
+  version: "1.0.0"
```
Comment (tmeschter) [inline plugin/skills/azure-iac-generator/SKILL.md]: You'll need to update this to "0.0.0-placeholder" and create a version.json file with the initial major.minor version. After that, versions numbers are updated automatically.

See other skills for examples.

File: plugin/skills/azure-iac-generator/references/auto-detection-rules.md (hunk 9 above)
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/references/auto-detection-rules.md]: good point. Added all options and if it cannot decide based on the current environment, it will default to `blob` and leave a comment.

File: plugin/skills/azure-iac-generator/references/procedures/diagram-parsing.md (hunk 10 above)
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/references/procedures/diagram-parsing.md:46]: added a default rule for `connects`

File: plugin/skills/azure-iac-generator/references/resource-configs-compute.md (hunk 11 above)
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/references/resource-configs-compute.md]: replaced the heading as suggested

File: plugin/skills/azure-iac-generator/SKILL.md (hunk 12 above)
Comment (Ba4bes) [reply plugin/skills/azure-iac-generator/SKILL.md]: Changed to placeholder and added the version file

**Review by jongio — COMMENTED** [review]
Addresses my previous feedback - all three items fixed (Storage PE groupIds, edge color fallback, Function App type key). Version handling looks correct now too. No new issues in the incremental diff.

**Review by tmeschter — DISMISSED** [review]

**Comment by Ba4bes** [comment]
@jongio @tmeschter tests should be fixed now

**Review by jongio — COMMENTED** [review]
Incremental check since my last review. The only new changes are a `vnet` trigger keyword addition to the snapshot and a test suite rename for clarity. No functional changes to the skill itself. Still looks good.

**Comment by tmeschter** [comment]
@Ba4bes Looks like the skills.json file was committed with the merge markers in place.

**Comment by Ba4bes** [comment]
@tmeschter  weird, I am sure I had removed them. Well they are gone now 👍

**Review by jongio — COMMENTED** [review]
Incremental check since my last review. The merge marker fix in tests/skills.json correctly includes both azure-iac-generator and azure-reliability. No issues with that change.

Two items from my earlier reviews remain open:

1. **Merge conflicts**: PR is currently in a conflicting state and can't merge. Needs a merge or rebase from main.
2. **Missing trailing newlines**: azure-authentication.md and bicep-parsing.md still lack a trailing newline (flagged in my first review).

The substantive code issues from prior rounds (Storage PE groupIds, edge color fallback, Function App type key) were all addressed. Skill structure, frontmatter, versioning, and test coverage look correct for this repo's conventions.

File: plugin/skills/azure-iac-generator/references/procedures/azure-authentication.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
+# Azure Authentication Check
```
Comment (jongio) [inline plugin/skills/azure-iac-generator/references/procedures/azure-authentication.md:1]: Still missing a trailing newline. Most editors and linters expect files to end with a newline character.

File: plugin/skills/azure-iac-generator/references/procedures/bicep-parsing.md [hunk 14]
```diff
@@ -0,0 +1,60 @@
+# Bicep Parsing Procedure
```
Comment (jongio) [inline plugin/skills/azure-iac-generator/references/procedures/bicep-parsing.md:1]: Same here: missing trailing newline.
